feat: Enable support for HTTP protocol to connect with Metastore

AnuragKDwivedi opened the pull request:
## Description
This PR introduces support for using HTTP(S) as the transport for the Thrift Metastore client. With this update, hive.metastore.uri can now accept HTTP(S) URLs to communicate with the metastore service.

Since Thrift APIs operate via POST requests, this change also enables authentication using HTTP Bearer Tokens or client secrets. These credentials must be configured to authenticate the metastore client with the metastore server.

Authentication can be set up using either:

Bearer Token: hive.metastore.http.client.bearer-token
Username & Secret: hive.metastore.http.client.plain.username and hive.metastore.http.client.plain.token
Additionally, this HTTP mode allows passing extra details, such as catalog name or account ID, as part of the request to the metastore.

## Motivation and Context
Thrift has supported HTTP-based transport for a long time. You can find the reference implementation here: [Thrift HTTP Client](https://github.com/apache/thrift/blob/master/lib/javame/src/org/apache/thrift/transport/THttpClient.java).

Additionally, Apache Hive introduced support for running the Hive Metastore server in HTTP mode, as documented in [HIVE-21456](https://issues.apache.org/jira/browse/HIVE-21456).

This PR updates HttpHiveMetastoreClientFactory to instantiate THttpClient as the transport mechanism when creating a Thrift Metastore client.

### **New Configurations:**  

New configuration properties have been introduced to enable connections to the metastore using the HTTP protocol:  

```
hive.metastore.http.client.tls.enabled=true  
hive.metastore.http.client.tls.truststore-path=/path-to-truststore.jks  
hive.metastore.http.client.tls.truststore-password=changeit  
hive.metastore.http.client.tls.keystore-path=/path-to-truststore.jks  
hive.metastore.http.client.tls.keystore-password=changeit  
```  

Authentication type default value set to **NONE**
If the authentication type is set to **BEARER**, the bearer token property must be configured:  

```
hive.metastore.http.client.authentication.type=BEARER  
hive.metastore.http.client.bearer-token=<bearer_token>  
```  

If the authentication type is set to **BASIC**, the username and password must be provided:  

```
hive.metastore.http.client.authentication.type=BASIC  
hive.metastore.http.client.auth.basic.username=<username>  
hive.metastore.http.client.auth.basic.password=<secret>  
```  

Read timeout default value set to 60 seconds and additional headers can be used to pass extra info to metatsore as a map.

```
hive.metastore.http.client.read-timeout
hive.metastore.http.client.additional-headers
```

Testing - 

<img width="675" alt="image" src="https://github.ibm.com/lakehouse/presto/assets/456943/9080036f-22e6-419b-b27f-8ef7d7180b66">
 

If release note is NOT required, use:

```
== NO RELEASE NOTE ==
```

## Summary by Sourcery

Add HTTP(S)-based Hive Metastore connectivity alongside existing Thrift support, with configuration, TLS/auth validation, and factory abstraction for selecting the appropriate client implementation.

New Features:
- Support connecting to the Hive Metastore over HTTP(S) via a new HTTP metastore client factory and configuration.
- Introduce configurable HTTP metastore client authentication (NONE, BASIC, BEARER) and additional request headers.
- Allow metastore URIs in configuration to use http/https schemes in addition to thrift.

Enhancements:
- Refactor metastore client creation behind a MetastoreClientFactory interface, separating Thrift and HTTP client factories.
- Update StaticHiveCluster to work with URI-based endpoints and improved error messages that include full metastore URIs.
- Add validation to prevent mixing HTTP(S) and Thrift or mixed HTTP/HTTPS schemes in metastore URI configuration.
- Generalize SSL context construction into a reusable HiveMetastoreUtils utility.

Build:
- Add HTTP client dependencies and a dedicated Maven test profile to run the HTTP-enabled Hive Metastore tests separately.

Tests:
- Add unit tests for HTTP Hive Metastore configuration parsing and defaults.
- Add mock HTTP/Thrift metastore client factories and extend existing cluster tests to cover HTTP mode and failure scenarios.
- Introduce an integration-style test that runs Hive in HTTP-enabled mode and verifies basic query and DDL/DML operations over HTTP.

steveburnett commented:
Please include documentation for the new configuration properties. https://github.com/prestodb/presto/blob/master/presto-docs/src/main/sphinx/connector/hive.rst might be a good place for them.

ZacBlanco commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/hms/http/HttpHiveMetastoreConfig.java:
```diff
@@ -0,0 +1,226 @@
+
+public class HttpHiveMetastoreConfig
+{
+    private HttpHiveMetastoreConfig.HttpHiveMetastoreClientAuthenticationType httpHiveMetastoreClientAuthenticationType = HttpHiveMetastoreConfig.HttpHiveMetastoreClientAuthenticationType.NONE;
```
```suggestion
    private HttpHiveMetastoreConfig.HttpHiveMetastoreClientAuthenticationType httpHiveMetastoreClientAuthenticationType = NONE;
```

ZacBlanco commented on presto-hive-metastore/src/test/java/com/facebook/presto/hive/metastore/hms/TestStaticHiveCluster.java:
```diff
@@ -46,6 +48,9 @@ public class TestStaticHiveCluster
             .setMetastoreUris("thrift://default:8080")
             .setMetastoreUsername("presto");
 
+    private static final StaticMetastoreConfig CONFIG_WIT_FALLBACK_AND_HTTP_MODE = new StaticMetastoreConfig()
```
```suggestion
    private static final StaticMetastoreConfig CONFIG_WITH_FALLBACK_AND_HTTP_MODE = new StaticMetastoreConfig()
```

ZacBlanco commented on presto-hive-metastore/src/test/java/com/facebook/presto/hive/metastore/hms/http/TestHttpHiveMetastoreConfig.java:
```diff
@@ -0,0 +1,78 @@
+                .put("hive.metastore.http.client.auth.basic.username", "lakehouse")
+                .put("hive.metastore.http.client.auth.basic.password", "test_password")
+                .put("hive.metastore.http.client.authentication.type", "BASIC")
+                .put("hive.metastore.http.client.additional-headers", "\"key\\\\:1:value\\\\,1, key\\\\,2:value\\\\:2\"")
```
Let's add more tests for different headers

ZacBlanco commented on presto-hive/src/test/java/com/facebook/presto/hive/TestHiveHttpEnabledMetastore.java:
```diff
@@ -0,0 +1,114 @@
+
+    private final Map<String, String> httpConfig;
+
+    TestHiveHttpEnabledMetastore() throws URISyntaxException {
```
```suggestion
    public TestHiveHttpEnabledMetastore() throws URISyntaxException {
```

agrawalreetika commented on presto-hive-common/src/main/java/com/facebook/presto/hive/CustomMetastoreAuthConfigUtils.java:
```diff
@@ -0,0 +1,119 @@
+        }
+        return token;
+    }
+}
```
@AnuragKDwivedi, I was wondering why we need this class in here?

AnuragKDwivedi commented on presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/hms/BridgingHiveMetastore.java:
Sure, I am keeping the files in same package for now. But this packages need to be reorganised. I will open another PR package changes.

AnuragKDwivedi commented:
[Update] - This PR depends on the Hive 4.0 upgrade in Presto. I noticed that one PR from @imjalpreet  is still pending merge  [prestodb/presto#24571](https://github.com/prestodb/presto/pull/24571)

I can also see that the Hive 4 Docker image has already been added to the Presto Docker repository. Once the Hive 4 upgrade is completed, I will update this PR and make it ready for review.
@faizdani

steveburnett commented:
Please edit the PR title to follow semantic commit style to pass the failing and required CI check. See the failure in the test for advice. If you can’t edit the PR title let us know and we can help.

AnuragKDwivedi commented:
> Please edit the PR title to follow semantic commit style to pass the failing and required CI check. See the failure in the test for advice. If you can’t edit the PR title let us know and we can help.

Changed branch name accordingly.

steveburnett commented:
> > Please edit the PR title to follow semantic commit style to pass the failing and required CI check. See the failure in the test for advice. If you can’t edit the PR title let us know and we can help.
> 
> Changed branch name accordingly.

[The required CI check is still failing](https://github.com/prestodb/presto/actions/runs/28008849668/job/82896994119?pr=24895). Look at the error in the check:
"Error: The PR title subject must start with a capital letter and not end with a period."

I think all you need to do is capitalize "Enable" in the PR title.

NivinCS commented:
Hi @AnuragKDwivedi - The dependent PR (https://github.com/prestodb/presto/pull/24571) has now been merged. Could you please move forward with opening this PR?

AnuragKDwivedi committed:
Enable HTTP protocol support in presto

This commit adds support for connecting to Hive Metastore via HTTP protocol.

Key changes:
- Added HttpHiveMetastoreClientFactory and HttpHiveMetastoreConfig
- Refactored metastore client factory architecture
- Added support for both Thrift and HTTP protocols
- Updated tests and documentation
- Renamed ThriftMetastoreModule to HmsModule for clarity

Signed-off-by: Anurag Dwivedi <anuragd916@gmail.com>

steveburnett commented on presto-docs/src/main/sphinx/connector/hive.rst:
```diff
@@ -154,7 +154,8 @@ Property Name                                            Description
                                                          fallback metastores. This property is required.
                                                          Example: ``thrift://192.0.2.3:9083`` or
-                                                         ``thrift://192.0.2.3:9083,thrift://192.0.2.4:9083``
+                                                        ``thrift://192.0.2.3:9083,thrift://192.0.2.4:9083`` or
```
This edit seems to have broken the table syntax, meaning the Hive Configuration Properties table does not compile. See this screenshot of a local doc build

<img width="621" height="814" alt="Image" src="https://github.com/user-attachments/assets/487b2a14-e3a6-4e73-a659-394fdc83316f" />

and compare it with https://prestodb.io/docs/current/connector/hive.html#hive-configuration-properties.

AnuragKDwivedi replied:
Resolved the indentation issue, and doc looking fine now
 
<img width="1366" height="996" alt="image" src="https://github.com/user-attachments/assets/251ea60f-0d15-4b7e-a059-5ee095d0ac41" />

steveburnett replied:
Looks good in a new local doc build for me, too. Thanks!

steveburnett reviewed (dismissed):
LGTM! (docs) 

Pull branch, local doc build, looks good. Thanks!

AnuragKDwivedi commented:
> LGTM! (docs)
> 
> Pull branch, local doc build, looks good. Thanks!

Thank you @steveburnett for the review.
